tools/java/org/unicode/cldr/api/PathMatcher.java

```diff
@@ -0,0 +1,276 @@
+     * any parent matchers from which this matcher might have been extended.
+     */
+    public boolean matches(CldrPath path) {
+        if (!locallyMatches(path)) {
+            return false;
+        }
+        if (!parent.isPresent()) {
+            return true;
+        }
+        while (path.getLength() > totalElementCount - elementMatchers.size()) {
+            path = path.getParent();
+        }
+        return parent.get().locallyMatches(path);
+    }
+
+    /**
+     * Returns whether this matcher matches a proper prefix the given path, including taking into
+     * account any parent matchers from which this matcher might have been extended. Note that the
+     * matcher for {@code "//ldml/foo/bar"} is not considered a prefix match for the path
+     * {@code //ldml/foo/barbaz}, since partial matching of an element name is not allowed.
+     */
+    public boolean matchesPrefixOf(CldrPath path) {
+        if (path.getLength() < totalElementCount) {
+            return false;
+        }
+        while (path.getLength() > totalElementCount) {
+            path = path.getParent();
+        }
+        return matches(path);
+    }
+
+    /**
+     * Returns whether this matcher matches the given path only in relation to the specification
+     * used to generate this matcher. This method is useful for nested visitation via {@link
+     * org.unicode.cldr.api.CldrData.PrefixVisitor PrefixVisitor} when path prefixes are visited
+     * before values.
+     *
+     * <p>This method assumed that any parent matchers from which this matcher was extended already
+     * match the path. The possible length of the parent matchers is taken into account however to
+     * ensure that the sub-sequence match for this matcher starts at the correct point in the path.
+     */
+    public boolean locallyMatches(CldrPath path) {
+        return (path.getLength() == totalElementCount)
+            && matchRegion(path, totalElementCount - elementMatchers.size());
+    }
+
+    private boolean matchRegion(CldrPath path, int offset) {
+        // offset is the path element corresponding the the "top most" element matcher, it
+        // must be in the range 0 ... (path.length() - elementMatchers.size()).
+        checkPositionIndex(offset, path.getLength() - elementMatchers.size());
+        // First jump over the path parents until we find the last matcher.
+        int matchPathLength = offset + elementMatchers.size();
+        while (path.getLength() > matchPathLength) {
+            path = path.getParent();
+        }
+        return matchForward(path, elementMatchers.size() - 1);
+    }
+
+    private boolean matchForward(CldrPath path, int matcherIndex) {
+        if (matcherIndex < 0) {
+            return true;
+        }
+        return matchForward(path.getParent(), matcherIndex - 1)
+            && elementMatchers.get(matcherIndex).test(path);
+    }
+
+    // --- Parsing of matcher path specifications ----
```

Comment:
FYI assuming this has been reviewed before, skipping ...

Reply:
Correct. This has been reviewed and is unchanged between the old and new versions of the class.